### PR TITLE
Express ESM versioned tests

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Added test ESM loader to properly mock out agent in versioned tests.
 * Added express ESM versioned tests.

## Links
Closes https://issues.newrelic.com/browse/NEWRELIC-3119

## Details
This will fail until we land the [test utils pr](https://github.com/newrelic/node-test-utilities/pull/141) and publish a new version.  In the meantime you can test this locally by checking out the branch in the test utils PR, running npm link in this branch and then just run ```npm run versioned:major```.  You will see the `express-esm` tests pass.  It's also worth noting ESM flavored versioned tests will have some limitations.  Since the loader only loads at the start of a test file, we cannot setup and teardown a mocked agent in a before/afterEach hook.  Going forward if we add more exclusive ESM versioned tests, the test will need to be made atomic to avoid cross-fire between tests.  The most notable changes in this PR are:

 * A test loader that swaps out a mock of the agent when the agent ESM loader tries to import the agent API(thanks @michaelgoin 🙏).
 * Refactoring of the test structure to be async in both loading the module under test and executing the test(in this case making a request and asserting info on transaction finished)
 * Saving a reference to the `transactionFinished` handler and removing after each test.  This was the biggest downside of not being able to bootstrap and teardown a new agent between every test.

When we originally scoped these port versioned tests to ESM we thought it'd be as easy as swap a few require statements for import and call it a day but it turned out to be a large effort.  That's why I did not port every single test from the cjs express tests.  Verifying both transaction and segment naming tests work is a good compromise and gives me enough confidence that the agent ESM loader is properly tested via integration tests.

### CODE REVIEW

File: test/.eslintrc.js [hunk 1]
```diff
@@ -15,5 +15,15 @@ module.exports = {
   },
   env: {
     mocha: true
-  }
+  },
+  overrides: [
+    {
+      files: ['*.mjs'],
+      rules: {
+        'node/no-unsupported-features/es-syntax': 'off',
```
Comment (bizob2828) [inline test/.eslintrc.js:23]: add TODO and link the jira ticket

File: test/.eslintrc.js [hunk 2]
```diff
@@ -15,5 +15,17 @@ module.exports = {
...
-  }
+  },
+  overrides: [
+    {
+      files: ['*.mjs'],
+      rules: {
+        // TODO: remove this when we decide on how to address
+        // here: https://issues.newrelic.com/browse/NEWRELIC-3321
+        'node/no-unsupported-features/es-syntax': 'off',
+        'node/no-unpublished-import': 'off',
```
Comment (bizob2828) [inline test/.eslintrc.js:26]: I will open PR to put this in the shared eslint ruleset. we already do it for the cjs equivalents: https://github.com/newrelic/eslint-config-newrelic/blob/main/index.js#L108

File: test/versioned/express-esm/package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "express-esm-tests",
+  "version": "0.0.0",
+  "type": "module",
+  "private": true,
+  "engines": {
+    "node": ">=14"
```
Comment (jmartin4563) [inline test/versioned/express-esm/package.json:7]: Do we need this `engines` if we're using the one inside of the `tests` stanza?

**Review by jmartin4563 — APPROVED** [review]
Just have one question about the test package.json, otherwise looks good!

File: test/versioned/express-esm/package.json (hunk 3 above)
Comment (bizob2828) [reply test/versioned/express-esm/package.json:7]: this one is for linting.  there's probably something specific in here but i can verify.  I just copied this from `test/versioned/express` so may not be necessary

**Review by jmartin4563 — APPROVED** [review]

**Pull request merged**